pkg/asset/installconfig/openstack/validation/platform.go

```diff
@@ -37,26 +38,78 @@ func ValidatePlatform(p *openstack.Platform, n *types.Networking, ci *CloudInfo)
-		if len(p.ExternalDNS) > 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("externalDNS"), p.ExternalDNS, "externalDNS is set, externalDNS is not supported when machinesSubnet is set"))
-		}
-		if ci.MachinesSubnet == nil {
-			allErrs = append(allErrs, field.NotFound(fldPath.Child("machinesSubnet"), p.MachinesSubnet))
-		} else if !validUUIDv4(p.MachinesSubnet) {
-			allErrs = append(allErrs, field.InternalError(fldPath.Child("machinesSubnet"), errors.New("invalid subnet ID")))
-		} else {
-			if n.MachineNetwork[0].CIDR.String() != ci.MachinesSubnet.CIDR {
-				allErrs = append(allErrs, field.InternalError(fldPath.Child("machinesSubnet"), fmt.Errorf("the first CIDR in machineNetwork, %s, doesn't match the CIDR of the machineSubnet, %s", n.MachineNetwork[0].CIDR.String(), ci.MachinesSubnet.CIDR)))
+// validateControlPlanePort validates the machines subnets and network, while enforcing proper byo subnets usage and returns a list of all validation errors.
+func validateControlPlanePort(p *openstack.Platform, n *types.Networking, ci *CloudInfo, fldPath *field.Path) (allErrs field.ErrorList) {
+	networkID := ""
+	hasIPv4Subnet := false
+	hasIPv6Subnet := false
+	if p.ControlPlanePort != nil {
+		if len(p.ControlPlanePort.FixedIPs) <= 2 {
+			if len(p.ExternalDNS) > 0 {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("externalDNS"), p.ExternalDNS, "externalDNS is set, externalDNS is not supported when ControlPlanePort is set"))
+			}
+			for _, fixedIP := range p.ControlPlanePort.FixedIPs {
+				if fixedIP.Subnet.ID != "" && !validUUIDv4(fixedIP.Subnet.ID) {
+					allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), errors.New("invalid subnet ID")))
+				}
+				subnet := getSubnet(ci.ControlPlanePortSubnets, fixedIP.Subnet.ID, fixedIP.Subnet.Name)
+				if subnet == nil {
+					subnetDetail := fixedIP.Subnet.ID
+					if subnetDetail == "" {
+						subnetDetail = fixedIP.Subnet.Name
+					}
+					allErrs = append(allErrs, field.NotFound(fldPath.Child("controlPlanePort").Child("fixedIPs"), subnetDetail))
+				} else {
+					if subnet.IPVersion == 4 {
+						hasIPv4Subnet = true
+					} else if subnet.IPVersion == 6 {
+						hasIPv6Subnet = true
+					}
+					if ok := isSubnetCIDRInMachineNetworkCIDR(n.MachineNetwork, subnet); !ok {
+						allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), errors.New("controlPlanePort CIDRs does not match machineNetwork")))
+					}
+					if networkID != "" && networkID != subnet.NetworkID {
+						allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), errors.New("fixedIPs subnets must be on the same Network")))
+					}
+					networkID = subnet.NetworkID
+				}
 			}
+			if !hasIPv4Subnet && hasIPv6Subnet {
+				allErrs = append(allErrs, field.InternalError(fldPath.Child("controlPlanePort").Child("fixedIPs"), errors.New("one IPv4 subnet must be specified")))
```

Comment:
In this case and for the one bellow I will leave `InternalError`  because `Invalid` requires to specify the details of the exact field that is invalid and in this case if the hole fixedIPs, which does not format really well in the error msg